core/trino-main/src/main/java/io/trino/operator/join/unspilled/JoinProbe.java

```diff
@@ -106,23 +96,81 @@ public Page getPage()
         return page;
     }
 
-    private boolean currentRowContainsNull()
+    private static long[] fillCache(
+            LookupSource lookupSource,
+            Page page,
+            Block probeHashBlock,
+            Page probePage)
     {
-        for (int i = 0; i < probePage.getChannelCount(); i++) {
-            if (probePage.getBlock(i).isNull(position)) {
-                return true;
+        int positionCount = page.getPositionCount();
+        List<Block> nullableBlocks = IntStream.range(0, probePage.getChannelCount())
+                .mapToObj(i -> probePage.getBlock(i))
+                .filter(Block::mayHaveNull)
+                .collect(toImmutableList());
+
+        long[] joinPositionCache = new long[positionCount];
+        if (!nullableBlocks.isEmpty()) {
+            Arrays.fill(joinPositionCache, -1);
+            boolean[] isNull = new boolean[positionCount];
+            int nonNullCount = getIsNull(nullableBlocks, positionCount, isNull);
```

Comment:
Why do we need to go though `boolean[] isNull` array as intermediate?
We can't we immediately compute `int[] positions` array (especially if there is a single nullable block)?

We can pre-allocate `int[] positions` to be `positions = new int[positionCount]`.

Reply:
We had discussion along similar lines here https://github.com/trinodb/trino/pull/13352#discussion_r976644690
The code which uses the array of positions subsequently takes exactly sized array as input parameter, so an array copy is needed to satisfy current API. Maybe we can change that, but let's leave it as a follow-up.

Reply:
I'm not sure follow-up will happen :). It is be better to have interfaces in more or less stable form form get go.

Going though `boolean[] isNull` seems like a waste.